### PR TITLE
Start

### PR DESCRIPTION
**Pull request by robertkowalski**

### CODE REVIEW

File: caron.js [hunk 1]
```diff
@@ -2,270 +2,253 @@
...
-    '    else',
-    '      redis.call(pushCmd, "bull:" .. jqueue .. ":paused", jobId)',
-    '    end',
-    '    redis.call("PUBLISH", "bull:" .. jqueue .. ":waiting@null", jobId)',
-    '  end',
-    'end'
-  ].join('\n')
-}
+    this.debug = opts.debug
+    if (this.debug) console.log('started')
```
Comment (f3rno) [inline caron.js]: Why not use the `debug` module?

File: caron.js [hunk 2]
```diff
@@ -2,270 +2,253 @@
...
-  lua: script,
-  numberOfKeys: 0
-})
+    const scripts = this.setupScripts(opts)
+    this.setupRedis(scripts)
 
-const elapsedTime = (start) => {
-  return process.hrtime(start)
-}
+    return this
```
Comment (f3rno) [inline caron.js]: This is unnecessary

File: caron.js [hunk 3]
```diff
@@ -2,270 +2,253 @@
...
+  }
+
+  stop (cb) {
+    if (!this.status.active) return
 
-const work = () => {
-  if (!STATUS.active || STATUS.processing) return
+    this.status.active = 0
```
Comment (f3rno) [inline caron.js]: Hehe :) Why not a boolean?

File: caron.js [hunk 4]
```diff
@@ -2,270 +2,253 @@
...
+    if (this.debug) console.log('stopping...')
 
-  if (rseed < STATUS.rseed) {
-    setTimeout(work, 5)
-    return
+    this.stopInter = setInterval(() => {
+      if (this.status.processing) return
+      if (this.debug) console.log('stopped')
+
+      clearInterval(this.stopInter)
```
Comment (f3rno) [inline caron.js]: Why not set a timeout?

File: caron.js [hunk 5]
```diff
@@ -2,270 +2,253 @@
...
-        let elapsed = elapsedTime(startTs)
-        console.log(res[1] + ' jobs processed in ' + elapsed[0] + 's,' + Math.round(elapsed[1] / 1000) + 'µs')
-      }
+    const rseed = Date.now()
 
-      setTimeout(work, program.freq || 25)
+    if (rseed < this.status.rseed) {
+      setTimeout(() => {
+        this.work()
+      }, 5)
```
Comment (f3rno) [inline caron.js]: You could set the timeout for `this.status.rseed - rseed` to avoid calling `work()` multiple times

File: caron.js [hunk 6]
```diff
@@ -2,270 +2,253 @@
...
+
+    if (debug) console.log(script)
+
+    this.redis.defineCommand('qwork', {
+      lua: script,
+      numberOfKeys: 0
+    })
+  }
+
+  setupScripts (program) {
```
Comment (f3rno) [inline caron.js:144]: Could be made static

File: caron.js [hunk 7]
```diff
@@ -2,270 +2,253 @@
...
+    const scripts = {
+      prefix: [
+        'math.randomseed(tonumber(ARGV[3]))',
```
Comment (f3rno) [inline caron.js:147]: Block strings would look great here :D Or breaking the lua out to its own file to get syntax highlighting.

File: caron.js [hunk 8]
```diff
@@ -2,270 +2,253 @@
...
+        '    end',
+        '    redis.call("PUBLISH", "bull:" .. jqueue .. ":waiting@null", jobId)',
+        '  end',
+        'end'
+      ].join('\n')
+    }
+
+    scripts.sidekiq = {
+      lua: scripts.ruby_common_1 + '\n' + [
+        'redis.call("LPUSH", "' + program.q_prefix + 'queue:" .. jqueue, payload)'
```
Comment (f3rno) [inline caron.js:241]: Hehe

File: test/integration.js [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+    })
+
+    testQueue.process((job, cb) => {
+      assert.equal(job.data.foo, 'bar')
+
+      cb(null)
+      testQueue.close().then(() => {
+        caron.stop(() => {
+          caron.redis.disconnect()
+          done()
```
Comment (f3rno) [inline test/integration.js:42]: 🍻

**Review by f3rno — CHANGES_REQUESTED** [review]

File: caron.js (hunk 4 above)
Comment (prdn) [reply caron.js]: @f3rno because you have too keep checking if caron stops processing, so you can safely shut down

File: caron.js (hunk 1 above)
Comment (prdn) [reply caron.js]: totally agree

File: caron.js (hunk 2 above)
Comment (prdn) [reply caron.js]: yeah, made for some benchmark, was lazy to remove

File: caron.js (hunk 3 above)
Comment (prdn) [reply caron.js]: indeed

File: caron.js (hunk 7 above)
Comment (robertkowalski) [reply caron.js:147]: aw right, maybe we can get this merged first - in the next pr i would put it into a seperate file.

**Review by f3rno — APPROVED** [review]
Nice start 👍

**Pull request merged**